### ISSUE TITLE
Do you guys want help with translations?

### ISSUE THREAD

**Issue by natanqb** [elishacloud/Silent-Hill-2-Enhancements#111]
Hi guys, first of all thank you very much for the amazing work you have done so far. I created this topic to see if you guys want help with translations to languages that aren't supported yet. If you want, i offer myself to help with a brazilian portuguese translation.

**Comment by Polymega** [comment]
Hi @natanqb,

Thanks for offering to help! At this time, we're not planning on adding any extra language options/translations that weren't already part of the PC version (English, French, German, Italian, and Spanish). 

However, I know there are other fans out there who are working (or have already done) other language packs for the PC version of the game. I know that Polish, Chinese, and Russian language packs exist out there. 

Anyway, this isn't something we're actively looking to do at this time, but there are definitely other fans out there who are working on such things and we think that's really awesome!

**Comment by Storm3000** [comment]
Hi, Polymega.

The spanish subtitles have some misspellings I want to fix. Is there any way to open these text files and modify it?

**Comment by Polymega** [comment]
@Storm3000 

Yes. Do the following: 

### BEFORE YOU BEGIN:
- **VERY IMPORTANT: ENSURE YOU INSTALLED THE EUROPEAN VERSION OF THE GAME!** All language files besides the English ones from the North American installation of the game have incomplete data/text! 

-------------------

### STEPS:
1. **Make a new folder** somewhere on your computer. These will house your edited game text files.
2. Go to `\Silent Hill 2\data\etc\message\` **and copy** all the Spanish text files to your new folder. These files will be named `*_s.mes`. Notice the "_s" at the end signifies they're the Spanish ones.
3. [Download the SH Resource Editor.](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/3118076/SH2.3.Resourse.Editor.-.For.Sounds.Text.zip)
4. Open the program and click the folder icon to load in a .mes file: 
![image](https://user-images.githubusercontent.com/27395260/56755124-c4ec8d80-675c-11e9-870d-9603b6c8a09f.png)
5. Click the "Decode" button:
![image](https://user-images.githubusercontent.com/27395260/56755163-dc2b7b00-675c-11e9-9953-9c5921929db0.png)
6. A list of all decoded text will populate: 
![image](https://user-images.githubusercontent.com/27395260/56755189-ed748780-675c-11e9-9830-8f02fd363a91.png)
7. Edit your text. Once edited, click "Insert current text." (If you do not click this and go to a different text field your changes won't be saved.):
![image](https://user-images.githubusercontent.com/27395260/56755252-1137cd80-675d-11e9-8816-78ec44b763d2.png)

-------------------

### WARNINGS:
- **WARNING:** Do not use unsupported characters! 
- **BIGGER WARNING:** Do not edit any field that pulls in live, dynamic text! It will break the code that pulls in the dynamic text! For example, do NOT alter the following: 
![image](https://user-images.githubusercontent.com/27395260/56755420-7db2cc80-675d-11e9-9e6e-ca93f0aa45c3.png)

In the image above, we see the text says, 
"The clock is stopped at : 
and 15 seconds."

In the game this could read as: 
"The clock is stopped at 8:47
and 15 seconds."

The game dynamically pulls in the time to populate this field, so do not touch this text field!!

Another example: 

![image](https://user-images.githubusercontent.com/27395260/56755635-f87be780-675d-11e9-9bae-f8e9249f9988.png)

This is the text for the save screen. In the image above, we see the text says, 
"Free space: KB"

In the game this could read as: 
"Free space: 627491 KB"

The game dynamically pulls in the storage space to populate this field, so do not touch this text field!!

-------------

### TESTING/FINAL NOTES:
- Once you finish editing a .mes file you need to click on "Replace temp -> original file" to save your changes: 
![image](https://user-images.githubusercontent.com/27395260/56755901-8952c300-675e-11e9-914d-b95bd0cb1c86.png)

- Once saved, place the new .mes file in `\Silent Hill 2\sh2e\etc\message\`
- **Always play test your changes!** Launch the game and trigger the new text. Ensure everything looks good and is still working. 
- Be careful not to make a single sentence line too long. If you change the word length, try to ensure it could still fit within the original 4:3 aspect ratio, adding in line breaks if they are needed:
![image](https://user-images.githubusercontent.com/27395260/56756043-e77fa600-675e-11e9-96d4-7aa60a009c65.png)

- Lastly, I'd recommend making a log file that documents all of your changes (before and after). That way, you'll always know what you changed/adjusted.

**Comment by Polymega** [comment]
@natanqb
I haven't tried it but, according to PC Gaming Wiki,
there exists a fan translation pack already made for Brazilian Portuguese:

https://pcgamingwiki.com/wiki/Silent_Hill_2:_Director's_Cut#Audio_settings

I'm not sure if it would be compatible with the high resolution font mod out of the box, though.

**Comment by IlDucci** [comment]
Good morning, I'd be willing to take the mantle of correcting and improving the official Spanish translation for the game.

I've already prepared the msg files and started doing edits, there's a few things I'd like to comment:
 - In the non-English versions of the ending letter, there are unknown characters at the beginning of each line. They're not identified by the program and the English version only has a "top" element, so I'm guessing these are control codes for centering these texts up and down in the scroll. You might want to update the tutorial so people don't touch them as well (I'd redo the tool to keep these in mind and do other things, but that's overkill for this case).
 - When you say the game dynamically edits the KB parts, does that mean the game actually goes to a specific byte or does it detect the KB part automatically? (Just being curious, I don't have to touch these fortunately).
 - Do you have any sort of save game packages in order to ease up the testing process?
 - Where should I upload any modified msg/txt files?
 - I talked in Twitter about the main menu's graphical texts, but I wanted to propose some additional changes that are more adaptational than anything:
     - Change "Opción" (Option) to the plural form, "Options". "Option" in singular seems to be a common Japanese trope/mistake, and there are more than one option available in that menu.
     - Change "Películas" (Movies) to "Vídeos". "Película/s" in Spanish is more oriented to feature films, rather than short movies.
     - Change "Nivel de Enigma" (Puzzle Level) to "Nivel de enigmas". Again, another plural form change.
     - Change "NACIDO DE UN DESEO" to "NACIDA DE UN DESEO". Supposedly, María's the one who is born here, if I'm not mistaken.

**Comment by IlDucci** [comment]
I'm going to reply myself about the KB size question:
 - The text files actually have variables within them (Thanks to the CMD text dumper, I was able to see those), but because they are not properly written in the GUI tool, they cause issues when misediting. If the variables could be seen in the GUI, people could edit those files more easily.

**Comment by Polymega** [comment]
Hi @IlDucci ,

Thanks for your interest in improving the experience for Spanish fluent players out there. 

> The text files actually have variables within them (Thanks to the CMD text dumper, I was able to see those), but because they are not properly written in the GUI tool, they cause issues when misediting.

I'm glad you figured out more information on the parts of text that pull in dynamic information.

> If the variables could be seen in the GUI, people could edit those files more easily.

The creator of this tool, John Modest, has gone completely radio silent many, many years back. Several people have tried contacting him to no success. I also believe the email on the programs we use is no longer valid, either. Mr. Modest also made other useful SH PC editing tools, such as a texture and audio explorer for the games. Unfortunately, I don't see these pieces of software ever being updated/improved upon. 

On that note: 

> In the non-English versions of the ending letter, there are unknown characters at the beginning of each line. They're not identified by the program and the English version only has a "top" element, so I'm guessing these are control codes for centering these texts up and down in the scroll.

That may very well be the case. I probably can't stress this part enough, but: Make sure you double-check (preferably triple-check) all edits you do in-game, to ensure the game is still displaying everything appropriately. 

> Do you have any sort of save game packages in order to ease up the testing process?

Sure, replace your `\Silent Hill 2\data\save\Folder 01\` folder with this one: [Folder 01.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/3374168/Folder.01.zip)

> Where should I upload any modified msg/txt files?

The `.mes` files are very small (usually only a few kilobytes apiece), so you can package them all up as a ZIP package and attach them to a reply post here. 

We're happy to include any improved text files in the next update release of Silent Hill 2: Enhanced Edition, but we're putting 100% trust in that you've play tested all your changes and the text is still displaying and behaving correctly for parts that you've changed. 

> I talked in Twitter about the main menu's graphical texts, but I wanted to propose some additional changes that are more adaptational than anything:

Okay, I can change these parts for you. I was the one who spoke with you on Twitter. Did you see my comment about "Historia Principal" still being too long? Any suggestions to shorten this with a different, more accurate translation? (Ignore the other words in the image below; this is before I apply your other translation changes.)

![D-_qW9yW4AgU1j2](https://user-images.githubusercontent.com/27395260/60910702-7fb20500-a24f-11e9-8ba7-2a22bab5f142.png)

**Comment by IlDucci** [comment]
Thank you very much for the saves, those will be very helpful.

About the MSG editors, well, guess I might not be that overkill when I talk about redoing the text tools. The text dumper shows most of the info, and I can keep working with these, but they lack the control code translations and, most importantly, they export certain elements in a way that breaks the character codification, so some variable areas would get broken when imported back.

I took a peek and figured out the basic format:
 - All pointers and control codes/variable codes are in Little Endian.
 - The pointers are 2-bytes per pointer, but the values are halved. This means all pointed texts must be aligned to two bytes.
 - First 0x02 bytes: Position of the last pointer. This can be used to identify the amount of pointers of each file.
 - Following 0x02 bytes: Pointer to a text string.
 - After that: Text strings. Padding elements seem to be written as 0x00.

I saw your comment on the Main Scenario line, and I didn't say anything because I was wondering. I thought the text "Carta desde el silencioso cielo" seemed wrong as well and I was thinking on the reasons why (Desde in this case doesn't convey the meaning of "from", but "written inside" in this particular case), so the answer would be: Changing the name to "Carta del silencioso cielo".

**Comment by Polymega** [comment]
Good news! Those new campaign titles fit. So the texture below will be the one released with the next update, which includes all your suggested translation changes. Thank you again for helping us with that: 

![image](https://user-images.githubusercontent.com/27395260/60923178-581d6580-a26c-11e9-93e0-4dc42b88b083.png)

> I took a peek and figured out the basic format:

Nice research! If you're down for it, you ought to stick around here as the more experienced help we have, the better!

**Comment by IlDucci** [comment]
Small thing I forgot: Please change OPCIÓNES with OPCIONES.

Thanks for the compliment, but I'm afraid my knowledge limits to pointer-finding, and other basic text stuff used in romhacking. I can't also program a damn thing, that's why I posted the format info.

**Comment by Polymega** [comment]
Sure thing, I've made the changes. :+1:

**Comment by Storm3000** [comment]
There's a little mistake on the spanish subtitles when James walks into the Heaven's Night main room. When he reads the neon lights "Heaven's Night", it says "Noche del Ciel". This is wrong. It should say "Noche del Cielo".

**Comment by IlDucci** [comment]
Thanks for pointing it out! I'll fix these.

A question for the dev team: Is it possible to add an additional language into the main game, or should it be prepped as an expansion set?

**Comment by Polymega** [comment]
> A question for the dev team: Is it possible to add an additional language into the main game, or should it be prepped as an expansion set?

I don't believe so. Fan translation packs work by replacing another language pack. For example: If you wanted to make a Dutch translation, you could save your files/overwrite the French `.mes` files. 

However, if you're simply improving the Spanish files you would keep your translation worked saved in the Spanish (`*_s.mes`) file format.

**Comment by IlDucci** [comment]
Darn. Here's the deal: The Spanish being used in Silent Hill 2 is Latin Neutral Spanish. I'm a Castilian Spanish translator, so I guess there can't be a double-language pack around. The differences between both are limited, but visible.

I'll try to contact a fellow who I worked with back when I did the retranslation of SH1, just so we can have both versions around. I guess Latin's should be the official.

**Comment by IlDucci** [comment]
Hello there, I'm sorry I didn't update you guys over the last 10 days. I've looked around and I've talked with a friend about doing a new MSG exporter/importer, the programming is on its way, so I won't touch the texts until we have a stable and working tool.

**Comment by Polymega** [comment]
> I've talked with a friend about doing a new MSG exporter/importer, the programming is on its way

That's super cool! Thank you and tell your friend thanks from all of us for making a new decoder/encoder.

**Comment by IlDucci** [comment]
Hi there. It's been a while, I'm sorry for the lack of information.

The tool's nearly complete. Right now it can extract properly all the contents from the PC version, XBOX version (Japanese included), and custom EE version files, and can reimport all of these versions, except Japanese (There's a bug, but it's left as low priority). The tool has some minor bugs, but I've been able to start doing some tests... and I think I'm going to start testing on the XBOX; as I have the feeling the font size/screen ratio constraints are higher there (And I might do a revision patch for XBOX and PS2).

In the meantime, I've found a couple of bugs, one of which, I believe, was left in the original PC game, but it doesn't appear on XBOX/PS2 files:

 - After loading in sequential order all the savegames, I think that around the hotel's elevator room, the Saved Games text at the tab, in the Save screen, gets sliced.
![sh2pc 2019-09-17 19-09-37-99](https://user-images.githubusercontent.com/5641681/65156498-7e5f3e80-da2f-11e9-802d-291616fdb641.jpg)

 - The Blood option selector is way to the left, not following the selection columns. In this case, I had to add extra spaces to the right of the option text to make this somewhat passable, but this behavior doesn't happen on XBOX.
![sh2pc 2019-09-18 16-08-45-49](https://user-images.githubusercontent.com/5641681/65156570-9fc02a80-da2f-11e9-8f53-18bd85239821.jpg)

**Comment by Polymega** [comment]
That's great to hear about the progress of your text editor! 

>  XBOX version (Japanese included) [...] except Japanese (There's a bug, but it's left as low priority)

I'm curious to how this will work with your editor. The PC version has NO Japanese support at all (it has the `*_j.mes` files, but I assume all data within is nulled). We tried importing the Xbox's Japanese text files and while some of the Japanese characters were restored, many of the characters were not. 

> After loading in sequential order all the savegames, I think that around the hotel's elevator room, the Saved Games text at the tab, in the Save screen, gets sliced.

Can you try replicating this bug again? Perhaps it's a specific save file that causes this once loaded? It would be nice if there's an easier way to replicate this, assuming we don't have to load a ton of other files in sequential order. 

> The Blood option selector is way to the left, not following the selection columns. In this case, I had to add extra spaces to the right of the option text to make this somewhat passable, but this behavior doesn't happen on XBOX.

This will be fixed in the next update, so you can undo your changes for this one on your end. More info about this bug is in this link: https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/129#issuecomment-501768243

**Comment by IlDucci** [comment]
> I'm curious to how this will work with your editor. The PC version has NO Japanese support at all (it has the `*_j.mes` files, but I assume all data within is nulled). We tried importing the Xbox's Japanese text files and while some of the Japanese characters were restored, many of the characters were not.

Surprisingly, the game has two methods of displaying text, one that mixes 1-byted and 2-byted characters (The Western one) and one that is strictly 2-byted characters (Japanese). This method is marked by adding an extra control code prior to each line in the Western versions. In our tool we just added a character table swap option, that will also add or avoid adding this extra control code. In order to display Japanese characters in a Western line, an extra E0 character is added prior to the Japanese character itself.

Yep, the Japanese files were dummied out for the PC version, filled with the fullwidth question mark.

About the characters, I did the character table nearly by hand (A large chunk of characters followed S-JIS, but some stray ones didn't). I tested pieces of the output (From the memos) with text transcriptions from Japanese websites and they matched 100%. I'd appreciate that someone who does know Japanese has a read at the dumped files, though.

> Can you try replicating this bug again? Perhaps it's a specific save file that causes this once loaded? It would be nice if there's an easier way to replicate this, assuming we don't have to load a ton of other files in sequential order.

Sure, I'll test this later on.

**Comment by Polymega** [comment]
That table swap feature sounds really nice. More language support is better for all fans of the world. :)

> I'd appreciate that someone who does know Japanese has a read at the dumped files, though.

When the time comes, I can try and make a call-out on our Twitter page to see if any Japanese fans/followers could help us.

**Comment by IlDucci** [comment]
I don't have any problem sharing the Japanese dump right now, but wouldn't it be... legally risky?

**Comment by Polymega** [comment]
What specifically would be dumped? Just `.mes` files?

**Comment by IlDucci** [comment]
Yeah, from mes to plain text.

**Comment by Polymega** [comment]
That should be fine. Sharing full game dumps is obviously a no-go, but message files should be fine.

**Comment by IlDucci** [comment]
Alrighty then. Serve yourself.

[messageJAPXBOX.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/3627613/messageJAPXBOX.zip)

**Comment by Polymega** [comment]
Perfect; thanks. Give me some time and I'll try and find someone fluent in Japan to check this all over and I'll report back to you. Thanks again!

**Comment by IlDucci** [comment]
I've just reached the hotel on my first playthrough and there's been no problems so far.

Also, I just realized the reason why the "Partidas g" text bug appear. It was a failure of my own, turns out that there was a bug with the comments function of the tool and the message file for that part of the hotel had one string less than what it should have. Seems like, besides the obvious disordering of lines, it can cause that sort of text corruption.

On the editing side, I've done quite some modifications, I even managed to swap the order of some multi-line strings (The first version of the Three Sins text) with no issues at all:

English:
`<CENTER-PARAGRAPH><COLOR-BLUE>I give you blood to atone 
	for the Three Sins.

<COLOR-BLUE> <COLOR-WHITE>... is written on top of the painting.

 Beneath the painting are three 
	 square depressions.<SEPARATORB>`

New Spanish:
`<COLOR-WHITE>Hay algo escrito sobre el cuadro.

<CENTER-PARAGRAPH><COLOR-BLUE>Te doy sangre para expiar
	los Tres Pecados.

<COLOR-WHITE>Bajo el cuadro hay tres agujeros
	cuadrados.<SEPARATORB>`

So the main problems are keeping the exact same number of strings and making sure the ending SEPARATORA/SEPARATORB/SELECT_YES_OR_NO are not removed.

**Comment by IlDucci** [comment]
I've finished both scenarios and I think the Spanish update is ready to go (Even if I didn't went through all the endings ingame, I crosschecked them with YT videos). I'm not going to call the friend I wanted to call about because I just found out he did some things I definitely don't like (Piracy with the Silent Hill 1 retranslation we worked on).

So I think I'm going to end with the texts. I think the only things that need a good research are the real actions that some variables do, but for that, we first have to release the tool.

Right now I've asked Infrid, the tool developer, to add a feature to the text files that includes a string number, in order to avoid the bug I've commented previously. He'd just have to add that, fix a couple of additional bugs, write the docs, and the tool should be made public.

**Comment by Polymega** [comment]
Hi @IlDucci ,

That's great to hear the tool is still coming along nicely! I also haven't forgotten about the Japanese verification thing; just give me some more time on that. 

Once you are 100% satisfied with the Spanish text updates let me know and we can include your work as part of the "Enhanced Edition Essential Files" download in the install guide. :+1:

**Comment by IlDucci** [comment]
Good morning everyone.

The tool is complete and released: https://gitlab.com/sh2msgconvert/sh2msgconvert

Please check it out. You guys probably know a bit (or lot) more about the control codes than we do.

As for the translation, I did some quick checks and I'm fine with its current state. I only have to see how to port it to PS2 to do a translation release on my website as well, but I guess we can start with your copy. The only question I have is if you'd like to keep the TXT files in the Github repo or if they'd be separate.

**Comment by Polymega** [comment]
Great work! It's always wonderful to have new and extra tools at our disposal for such work, so we all appreciate the software you've made. :D Thanks again to you and your friend!

> As for the translation, I did some quick checks and I'm fine with its current state. I only have to see how to port it to PS2 to do a translation release on my website as well, but I guess we can start with your copy.

No worries. Once you're 100% satisfied with the edits (and play tested all the changes) just attach a ZIP package of all the new `*_s.mes` files here and we'll include them in the next update release for the project. Thank you as well for helping to correct sentence/grammar issues with the Spanish translation of the game.

> The only question I have is if you'd like to keep the TXT files in the Github repo or if they'd be separate.

Apologies, as I may be misunderstanding your statement here, but for release with our project, your translation work will be included in the "[Enhanced Edition Essential Files](http://enhanced.townofsilenthill.com/SH2/index.htm#installEssentialFiles)" download as the game-ready `.mes` format. (This particular download package isn't part of a Git repository.)

**Comment by IlDucci** [comment]
> > 
> 
> Apologies, as I may be misunderstanding your statement here, but for release with our project, your translation work will be included in the "[Enhanced Edition Essential Files](http://enhanced.townofsilenthill.com/SH2/index.htm#installEssentialFiles)" download as the game-ready `.mes` format. (This particular download package isn't part of a Git repository.)

Sorry, I thought you'd be adding the .txt files as sources for the repository. Understood.

**Comment by Polymega** [comment]
I see. I have no say in that, but I'd imagine Elisha probably wouldn't add the .txt files to this repo as the tool to work with them isn't part of this project.

Could the .txt files be added to the [sh2mgsconvert GitLab repo](https://gitlab.com/sh2msgconvert/sh2msgconvert)? Perhaps we could link to that repo to let people know about it, in case they want to edit their own .mes files?

**Comment by Polymega** [comment]
I'm sitting down with the software and have run into an issue. I receive the following error: 

`[ERROR] [WinError 2] The system cannot find the file specified: 'stage_town_west_msg_e.txt'`

I've placed `stage_town_west_msg_e.mes` into the same folder as `sh2msg.exe` and am running the command prompt from the folder location, then typing the following: 

`sh2msg stage_town_west_msg_e.mes`

**Comment by IlDucci** [comment]
Just checked it out and the EXE fails when exporting any asset, on my case, with the `[ERROR] argument should be string, bytes or integer, not WindowsPath` error.

Seems like this is an issue on the compiles made by Gitlab. I've already told that to Infrid, in the meantime, please try the non-compiled py scripts instead.

TXT files should not be blank, they always must have the header commented in the README file.

**Comment by Infrid** [comment]
Hello!

I just fixed it, you can find the new binaries on gitlab. The problem was related to the automatic build process, updating the python version and few lines in the scripts solved the issue.

**Comment by Polymega** [comment]
Hi @Infrid !

I tried the updated build on my Win10 laptop and it works great! Wonderful work to you and @IlDucci once more!

I had to use `.\sh2msg *.mes` instead of `sh2msg *.mes` to get it to work. Just mentioning for any future troubleshooting.

Also, for future troubleshooting, on my Win7 desktop I keep receiving the following error: 

`Error loading Python DLL 'C:\Users\<username>\AppData\Local\Temp\_MEI51122\python37.dll'.`

I installed [the latest version of Python (3.8.0)](https://www.python.org/downloads/) and have this DLL located under `C:\Users\<username>\AppData\Local\Programs\Python\Python<version-#>\python37.dll`

When I copied this DLL to `\AppData\Local\Temp\_MEI51122\` the same error would occur but the folder `_MEI51122` would change to something different like `_MEI51962`.

I'm not sure if that's a Win7 specific issue or not, but wanted to bring it to your attention. It's also worth noting I regularly delete/clear out my `\AppData\Local\Temp\` folder to gain gigabytes worth of disk space back.

**Comment by Polymega** [comment]
By the way guys, I just want to say again how tickled pink I am with this new software.

I am now able to correctly fix the Born From a Wish warning message, in terms of alignment. [Before, I had to do a very sloppy job in order to fix it.](https://github.com/ThirteenAG/WidescreenFixesPack/issues/620)

Now, thanks to your new software, I can properly edit `stage_x_heaven_msg_*.mes`:

```
<TEXT-POS-120>This additional scenario has been
	written as a supplement to the main
	game. We recommend that it be played
	only after completing the main game.
	
	Do you wish to play?
```

Originally, this was `<TEXT-POS-040>`, and my previous "hack" to fix it was aligning the text to the top and then line breaking a bunch of times down. 

So great work again!!

@IlDucci 
I've went ahead and fixed this for all `stage_x_heaven_msg_*.mes` files but you may want to also apply this change to your revised Spanish `.mes` project as well.

**Comment by IlDucci** [comment]
> ```
> 
> ```
> 
> 
> Originally, this was `<TEXT-POS-040>`, and my previous "hack" to fix it was aligning the text to the top and then line breaking a bunch of times down.

According to the dumps I've made of the original files, both PC and XBOX (and PS2, I've already done the dump and port, just needs testing), only your file has the <TEXT-POS-XXX> variable. You guys might want to use a diff check on all your changes, because that would be the second time some control codes are off in contrast with the original files (The other one was a missing variable in common or option).

**Comment by Polymega** [comment]
> EDIT: I just saw that you added that positioning initially, not that it was from the original texts. I've seen the issue with Spanish and I've added the control code to my PC copies.

Sorry, I should've clarified that. Yes, it's an inherent bug for the PC version and adding the `<TEXT-POS-###>` can properly fix it (as opposed to my earlier, jankier methods of trying to fix it).

> (The other one was a missing variable in common or option).

Thanks, I'll double-check options and common. 👍 

Also, I haven't forgotten about the Xbox Japanese files to proof read. [I've put out a message](https://twitter.com/WhereAllBegins/status/1203045310458736645) in hopes that someone native in Japanese will respond.

**Comment by Infrid** [comment]
@Polymega I fixed that nasty bug on windows platform!

**Comment by Polymega** [comment]
Thanks @Infrid!

I still receive the error on my desktop, unfortunately. 

![image](https://user-images.githubusercontent.com/27395260/70669526-a6cc9280-1c44-11ea-8371-35a6139e9d98.png)

After a little more digging it turns out this is indeed a < Win10 issue:

- https://github.com/pyinstaller/pyinstaller/issues/4345#issuecomment-535230118
- https://pyinstaller.readthedocs.io/en/stable/usage.html#windows

With Win7 support officially ending in January 2020, I wouldn't worry about trying to make this compatible with < Win10 OS. I'm perfectly fine to make edits on my Win10 laptop and transfer the files over to my desktop.

Thanks again though to you and IlDucci for your fantastic work with this program--it's incredibly useful!

**Comment by IlDucci** [comment]
I've finished doing the tests in PS2 for the Castilian Spanish version. This helped me change a few things on the "Neutral" version, so I consider the texts completed. The only comment was that the game was stuttering and had load issues starting in one day in particular, but turns out it's because of the media I was using (They were supposed to be Taiyo Yuden discs, but they seem to be the T03 bad batch).

Anyway, to add the cherry on top of the cake, I've researched the graphics and I have translated the main menu and Konami logos (for crediting) on PS2 and XBOX (By using Console Texture Explorer and some trickery)... and then I saw what happened on PC. My God, why the hell did they have to pick DXT2 for the transparent contents? I'd like to suggest if you guys can change the game code to expect a DXT3 texture (As trying to use an existing DXT3 causes decoding issues) so the main menu looks much better (DXT2 is a horrible compression, and DXT3, for the main menu (Grayscale with 256 color alpha) is a much better choice. Hell, DXT3 is a better choice than DXT2 for nearly everything.

**Comment by Polymega** [comment]
Nice notes! I'm not sure if SH2 PC itself uses DXT2 or if the [SH Texture Explorer](http://alchemillahospital.net/how-to-use-sh-texture-explorer/) uses DXT2.

I do know the butterfly texture [used a 1-bit alpha channel which we've addressed.](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/49#issuecomment-474169114)

In any case, whenever Aero_ and I do any texture edits with transparency we always save them out as DXT3 .DDS format and manually swap the hex data in the `.map` files, so no worries there. (Although, back in the day, we used to use SH Texture Explorer before we started manually changing the data ourselves.)

> Also, if you want to use my PSDs, here's a sample of how my menu texts:

Thanks! [We've done translations for the main menu text, which will be included in the next update.](https://twitter.com/WhereAllBegins/status/1146238025912999936)

However, we're not too worried about making the main menu font look perfect at the moment, because [we plan make a high definition, properly-fitted main menu + text](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/48#issuecomment-464410040) in the future. Once we make that, we'll be sure to make the main menu text look as good as it can be.

> This helped me change a few things on the "Neutral" version, so I consider the texts completed.

Congratulations! As mentioned before, if you'd like me to include your text improvements into the project for the next update, be sure to attach them here whenever you're ready. I already have your's and @Infrid's names included in the next update for the website and PC Credits movie under the 'Special Thanks' section. ❤️ 

Also, I haven't had anyone respond to me yet about proof-reading the Japanese `.mes` files...

**Comment by IlDucci** [comment]
> In any case, whenever Aero_ and I do any texture edits with transparency we always save them out as DXT3 .DDS format and manually swap the hex data in the `.map` files, so no worries there. (Although, back in the day, we used to use SH Texture Explorer before we started manually changing the data ourselves.)

Maybe the Texture Explorer could use a revamp, but I don't think we're going to spend time there, I'm afraid.

> However, we're not too worried about making the main menu font look perfect at the moment,

The problem isn't just at the font choice. It seems that the outer glow was kinda smashed and broken for the PC version. If you plan to do a HD replica, don't use the PC version's texture as a reference, but rather go with the XBOX equivalent (It's on 8bpp mode, which is closer to 1:1 with PS2 than any DXT compression).

> Congratulations! As mentioned before, if you'd like me to include your text improvements into the project for the next update, be sure to attach them here whenever you're ready. I already have your's and @Infrid's names included in the next update for the website and PC Credits movie under the 'Special Thanks' section. heart

Here's the text files for the "Neutral" version, the one you will be interested. I would like to request if you can add some sort of comment for the alternative, Castilian version on the website once I release the patch (Dec. 28's going to be).

[messageESN.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/3965453/messageESN.zip)

And here's a quickie couple of changes for the Pause txt file:

[exe_str.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/3965454/exe_str.zip)

> Also, I haven't had anyone respond to me yet about proof-reading the Japanese `.mes` files...

Damn, that's too bad. I guess we'll have to wait, either until someone reaches you guys or until someone complains about the tool. Heh.

**Comment by Polymega** [comment]
> Maybe the Texture Explorer could use a revamp, but I don't think we're going to spend time there, I'm afraid.

No worries. For me personally, nowadays, I only use SH Texture Explorer to preview the texture contents of a `.map` file. Once I find the `.map` file I'm looking for, I will manually hex edit the file to extract and re-insert any edited texture work.

> Here's the text files for the "Neutral" version [...] And here's a quickie couple of changes for the Pause txt file:

Perfect! These will be added/included with the next update release. Thank you!

> I would like to request if you can add some sort of comment for the alternative, Castilian version on the website once I release the patch (Dec. 28's going to be).

Sure thing. I'm not sure when I can include alternate language packs on the site, but that **_is_** something I ultimately want to do. I want fans of all different languages to enjoy the game and our project.

The two things that are currently preventing me from including additional fan language packs on the site are: 1) I need to decide where/how to include them on the site that "flows" correctly, and, more importantly, 2) all language packs need to be play tested to ensure everything is working right, especially with or without using the `UseCustomFonts` mod. (I know your Castilian version will be good to go as-is.)

Let me think on how I should approach this for now, but  I fully intend to include additional language packs which can happily include the Castilian version. In the meantime, here is a download list of other fan language packs, of which it can also be added here: https://www.pcgamingwiki.com/wiki/Silent_Hill_2:_Director's_Cut#Localizations

**Comment by elishacloud** [comment]
> I'm not sure if SH2 PC itself uses DXT2 or if the [SH Texture Explorer](http://alchemillahospital.net/how-to-use-sh-texture-explorer/) uses DXT2.

I just ran a few tests and SH2 uses a combination of DXT1, DXT3, DXT4 and DXT5.  Most being DXT5, as far as I can tell.

**Comment by Infrid** [comment]
@Polymega I tested the binaries against an up to date Win7 installation, it might be the case where we need some weird update installed on the system in order to run the tool. Eventually the user can always install the Python interpreter, it's not the best user-friendly solutions, but it works. I will try to find a solution when I have some spare time.

Anyway looking forward for the release! :)

**Comment by Polymega** [comment]
Thanks again, @Infrid!

If the issue is what I linked to earlier, a possible easy fix, according to another Python developer, [would be to include the needed CRT DLLs](https://github.com/pyinstaller/pyinstaller/issues/4345#issuecomment-535230118) inside the same directory as the binary for its download package.

I tried figuring out exactly what those DLLs are, and where to safely obtain them, but didn't have any luck on that front. [This page](https://pyinstaller.readthedocs.io/en/stable/usage.html#windows) just says they're "Visual C++ run-time .dlls."

**Comment by IlDucci** [comment]
> **Edit:** Oh, and @IlDucci I've asked my friend to make a call-out for a Japanese proof-reader as well. Hopefully something will come of it:
> https://twitter.com/SH_Historical/status/1206635946386444288
> https://www.facebook.com/silenthillhistoricalsociety/posts/2748899655156865

Let's hope that attracts someone interested.

In the meantime, I was browsing around the issues and saw your things with Russian. This tool could work with any non-EFIGS and Japanese languages if you can research out the character table and edit the table.txt file (To play around translations and tables, I'd suggest using the uncompiled py scripts, as the tool right now doesn't support "outside" tables).

**Comment by Polymega** [comment]
@IlDucci 

> Also, I just realized the reason why the "Partidas g" text bug appear. It was a failure of my own, turns out that there was a bug with the comments function of the tool and the message file for that part of the hotel had one string less than what it should have. Seems like, besides the obvious disordering of lines, it can cause that sort of text corruption.

Can you explain this part in more detail? I switched over to the new Spanish pack and am also experiencing this issue now. It doesn't happen for the "Load Game" selection menu, but it does happen for the "Save Game" selection menu.

![sh2pc 2019-12-23 21-19-47](https://user-images.githubusercontent.com/27395260/71389432-5a7d3d00-25ca-11ea-85c1-13b745da7961.png)

Can you also test this to see if you experience it as well? Load a game file, then select "Save Game" and scroll the save files.

**Comment by IlDucci** [comment]
> 
> 
> **Edit:** I used the vanilla/default Director's Cut `.mes` files and the save folder tab's text seems to be glitchy here, as well. It looks like this may be an inherent bug with the game, but it only happens on the "Save Game" screen?

I thought it was related to a mishap I caused, I accidentally deleted one string from a file in this particular screen and I thought the letters being removed was something related to memory issues, (that's why we added a string number to the file, as a precaution) but if you're telling me that this happened with the vanilla game... Then yes, seems like this happens just here.

Seems like there's a limit there to exactly 11 characters, just the ones needed for English (Saved Games).

**Comment by Polymega** [comment]
> Seems like there's a limit there to exactly 11 characters, just the ones needed for English (Saved Games).

I tested this the other day and that may not be the case. At one point it went down to just "Partidas":

![image](https://user-images.githubusercontent.com/27395260/71424997-f4de8e80-2665-11ea-804f-89a6403fe11e.png)

> There's also another thing that bothers me. Did you guys managed to make the Save/Load screen display a date, rather than the ingame elapsed time?

Yes, press right or left on the game file selection screen, it will cycle through different time display options.

**Comment by Polymega** [comment]
I can't look into this too much today or tomorrow, but it seems to be a result of using special characters (possibly such as ó or ª) for `m_card(_x)_msg_*.mes`. I used the English `.mes` wording, but changed the tab name to "Partidas guardadas", saved it out as `m_card_x_msg_s.mes`, and the tab words never glitched out:

![image](https://user-images.githubusercontent.com/27395260/71425089-0d02dd80-2667-11ea-8bbc-195b2d43cc8c.png)

**Comment by IlDucci** [comment]
> > There's also another thing that bothers me. Did you guys managed to make the Save/Load screen display a date, rather than the ingame elapsed time?
> 
> Yes, press right or left on the game file selection screen, it will cycle through different time display options.

Well, crap. That means I translated the save locations without keeping those dates in mind. As a matter of fact, the PC version is the most restrictive here (The longest choice is the date, which is displayed with both the date and the time, while in XBOX and PS2 they are split between two choices, one for the date and one for the time). I have to abbreviate a couple of names. (MAKE SURE YOU TELL THIS TO ANY PROSPECTIVE TRANSLATORS and check these! It's these kind of semihidden things that can make a good translation not that good).

As for the Saved Games slicing, I'm not so sure it's an accent thing. I have a feeling it's a overall/allocated size thing, as the PC version's m_card_x file is the longest one in all consoles (Because of all the extra save location names).

I'm gonna do a quick test removing all accents from the Spanish version.

**Comment by IlDucci** [comment]
It's an allocated size thing. I've done two things:
 - First I removed every special character from the Spanish m_card_x file. The behavior was still in place when trying to save from a save point (Going Pause -> Load Game didn't display this behavior):
 - After that, I took the m_card_x file and removed every XBOX and PS2 related string. Look at the results:

![sh2pc 2019-12-24 23-40-35](https://user-images.githubusercontent.com/5641681/71426235-eb294b00-26a6-11ea-9c1d-b5cef34b4e8f.png)

**Comment by Polymega** [comment]
Funny timing, as I just figured out another solution, I think:

`Presiona Retroceso para borrar archivos.`

This string is too long and is causing the sliced/missing text elsewhere. When I shortened it, the issues no longer happened.

**Comment by IlDucci** [comment]
You were right. Damn... Do you want me to shorten this? This might also affect other official languages.

**Comment by Polymega** [comment]
I just edited my post. My OCD really wants to keep playing with this right now but I have to head out for  (Merry Christmas and Happy Holidays!). Here's what I edited into my last post: 

The tab text would "flicker" for a moment when scrolling through save file selections, but the full verbiage remained in my tests. So this may not be a definite solution. But it does seem like it may be an overall text-length issue for the .mes file?

**Comment by IlDucci** [comment]
Either that or, if your particular line makes any sense, some sort of text-related buffer, specifically, text that is printed on space could have some sort of limitation?

I just checked the other official languages. French and German also have a long Backspace string. So...

**Comment by Polymega** [comment]
Yeah, the other official languages also experienced glitchiness, to varying degrees, for the tab text in my tests. This must have been overlooked during the PC conversion process.

**Comment by IlDucci** [comment]
[sh2-esn-11.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/3999167/sh2-esn-11.zip)

Here's the new version of the Spanish texts, with the offending save locations shortened.

**Comment by Polymega** [comment]
@IlDucci 
I think I got it figured out. There are two things you must do:
1. The "delete with backspace" line _must_ be short in length. This is probably contextually incorrect but, for the Spanish version, I changed it to "Retroceso para eliminar".
2. The tab text _must_ be eleven characters or less.

**If you shorten the "backspace" line but make the tab text longer than 11 characters:**
The full tab text verbiage will remain intact but will momentarily flicker when scrolling through save files.

**If the "backspace" line is long and the tab text is shorter that 11 characters:**
The full tab text verbiage will remain intact but will momentarily flicker when scrolling through save files.

**If the "backspace" line is long and the tab text is longer that 11 characters:**
It can slice the tab text (the current issue we're having).

I haven't figured out the maximum number of characters you can use for the "backspace" line.

These changes _seem_ to fully fix the issue: 

![image](https://user-images.githubusercontent.com/27395260/71450864-085a2a00-2738-11ea-8109-20883d3ab19b.png)

... so the question is: Is "Retroceso para eliminar" an acceptable substitute?

**Comment by IlDucci** [comment]
Is acceptable, but I'd like to polish things a bit further:
 - The tab text should be "Archivos" or "Guardados" (I'd stick with the first one), in plural.
 - The Backspace line can be made a bit more snappier with this: "Retroceso: Eliminar".

You can do the changes in this case, though I'll do them as well in my local files for the Dec. 28 release. In any case, can you guys do something about it via code?

**Comment by Polymega** [comment]
Awesome, thanks again IlDucci. I've adjusted the coloring on some of the save file names. Check out these modified files for your records. Basically all save files will be white, and "Game Result" & "Quick Save" will be yellow. New save/Saving/Deleting will be blue. I've edited this across the board for all language files.

[m_card_msg_s.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/4003625/m_card_msg_s.zip)

> In any case, can you guys do something about it via code?

Possibly. But, if it can be "fixed" by shortening two strings, it's easier to try and keep those lines short, instead. Saves time and resources for other tasks.

**Comment by IlDucci** [comment]
Well, as long as you don't forget about that for a later time, I'm okay with that.

I have released the patch on my site, but I've also did some final edits on the eleventh hour. There was a typo I didn't saw on the Acacia file, and I've did a couple of neutralization tweaks in the observatory.

[sh2e.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/4007121/sh2e.zip)

I've posted my project here: https://tiovictor.romhackhispano.org/silent-hill-2-directors-cut/

And the news update is here: https://tiovictor.romhackhispano.org/2019/12/28/hemos-llegado-100-textos-traducidos/ (I've mixed it with an April Fools joke, as in Spain it's celebrated in December 28th, just in case you see something funny).

**Comment by Polymega** [comment]
Thanks again, IlDucci. I'll also make an announcement on our Twitter page for your Castillian language pack with a link to your project page for it in the near future. We really appreciate all your help to make the user experience for Spanish-speaking fans better.

**Comment by IlDucci** [comment]
I've done a minor tweak on the Options menu, after some suggestions.

[option_msg_s.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/4009820/option_msg_s.zip)

**Comment by Polymega** [comment]
Got it; thanks.

**Comment by Polymega** [comment]
@IlDucci @Bruno3589
Can you guys give me a quick translation of the following line in your respective languages? This is for a save file name: 

### Toilet/Bar/Motorhome

@IlDucci 
You will want to update that decription for your Castilian project as well. Basically, if you custom save at the beginning bathroom, BAR Neely's, or the Motorhome, the save file currently just says "Toilet."

**Comment by IlDucci** [comment]
Thanks for the heads-up! I'd use this:

`Servicios/Bar/Caravana`

**Comment by Bruno3589** [comment]
Here's for Italian:
`Gabinetto/Bar/Roulotte`

**Comment by Polymega** [comment]
@IlDucci 
In reference to your post here: https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/111#issuecomment-532780215

Bigmanjapan has a Japanese contact who read over the extracted Xbox Japanese .mes files and has confirmed the extraction is good. Besides any inherent typos (that have always been in the Japanese text), everything was extracted without issue.

I'm not sure if you'd still want to restore the Japanese text for SH2 PC, but if it's something you'd like to look into, you _might_ need to disable the HD font (`UseCustomFonts = 0`) as we never made HD versions of the Japanese characters (so as to keep the HD font file's size down).

**Comment by IlDucci** [comment]
Well, my only intentions with the Japanese table was to make sure our tool was 100% usable on every existing official language. Now that your contact (can you tell us its nickname, so we can credit him?) has confirmed that the Japanese language can be dumped and edited, I can consider the tool finished.

**Comment by Polymega** [comment]
Understood and no problem. I'll ask BMJ his contact's name a little later and let you know. Thank you both again, IlDucci and Infrid, for all your help with this.

**Comment by Polymega** [comment]
Thank you all for your hard work in making better text editing tools and fixing typos/errors in your respective languages.

**Comment by kayforbe** [comment]
i can't open the EditSH23Msg program

**Comment by JohnModest** [comment]
> The creator of this tool, John Modest, has gone completely radio silent many, many years back. Several people have tried contacting him to no success. I also believe the email on the programs we use is no longer valid, either. Mr. Modest also made other useful SH PC editing tools, such as a texture and audio explorer for the games. Unfortunately, I don't see these pieces of software ever being updated/improved upon.

Yes.  To edit text, I made another program a long time ago. The source is not left.

**Comment by JohnModest** [comment]
shtxtconv.exe   mes -> txt   txt->mes

**Comment by Polymega** [comment]
Oh wow! What a surprise! Nice to see you here, John. 

A lot of progress has been made in the Silent Hill modding scene over the years. 

Two great examples include a new .mes <-> .txt editing tool called [SH2MSGConvert](https://github.com/loadwordteam/SH2MSGConvert). In addition to this, [a SH2 level editor](https://github.com/pmttavara/ph2) is currently being developed to allow us to modify the game's assets.
